File: lc/coreprep/implementation/patterns_reference_implementation/modified_binary_search/minimum_difference_element.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// O(logn) / O(1)
int find_min_diff_pos(const vector<int> &a, int low, int high, int k) {
  while (low <= high) {
    int mid = low + (high - low) / 2;

    if (a[mid] < k) {
      low = mid + 1;
    } else if (a[mid] > k) {
      high = mid - 1;
    } else {
      return a[mid];
    }
  }

  return (abs(k - a[low]) < abs(k - a[high])) ? a[low] : a[high];
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k;
  cin >> tt;

  while (tt--) {
    cin >> n >> k;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    cout << find_min_diff_pos(a, 0, n - 1, k) << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/MinimumDifferenceElement.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class MinimumDifferenceElement {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        System.out.println(minDiffElement(a, 0, n - 1, k));
      }
    }
  }

  // O(logn) / O(1)
  public static int minDiffElement(int[] a, int low, int high, int k) {
    if (k < a[low]) {
      return a[low];
    }

    if (k > a[high]) {
      return a[high];
    }

    while (low <= high) {
      int mid = low + (high - low) / 2;

      if (a[mid] < k) {
        low = mid + 1;
      } else if (a[mid] > k) {
        high = mid - 1;
      } else {
        return a[mid];
      }
    }

    if (Math.abs(k - a[low]) < Math.abs(k - a[high])) {
      return a[low];
    }
    return a[high];
  }
}